distributed/client.py

```diff
@@ -108,42 +111,75 @@
 
 logger = logging.getLogger(__name__)
 
-_global_clients: weakref.WeakValueDictionary[
-    int, Client
-] = weakref.WeakValueDictionary()
-_global_client_index = [0]
-
 _current_client = ContextVar("_current_client", default=None)
 
 DEFAULT_EXTENSIONS = [PubSubClientExtension]
 # Placeholder used in the get_dataset function(s)
 NO_DEFAULT_PLACEHOLDER = "_no_default_"
 
 
-def _get_global_client() -> Client | None:
-    L = sorted(list(_global_clients), reverse=True)
-    for k in L:
-        c = _global_clients[k]
-        if c.status != "closed":
-            return c
-        else:
-            del _global_clients[k]
-    return None
+class _GlobalClientManager:
+    def __init__(self):
+        self._lock = threading.RLock()
+        self._global_clients: weakref.WeakValueDictionary[
+            int, Client
+        ] = weakref.WeakValueDictionary()
+        self._global_client_index = 0
+        self._set = None
+
+    def _get_global_client(self) -> Client | None:
+        with self._lock:
+            L = sorted(self._global_clients, reverse=True)
+            for k in L:
+                c = self._global_clients[k]
+                if c.status != "closed":
```

Comment:
can the client be in the "closed" status here? - `client.close` calls `_del_global_client`